lls/arp.c

```diff
@@ -178,6 +183,10 @@ process_arp(struct lls_config *lls_conf, struct gatekeeper_if *iface,
 		ether_addr_copy(&iface->eth_addr, &arp_hdr->arp_data.arp_sha);
 		arp_hdr->arp_data.arp_sip = iface->ip4_addr.s_addr;
 
+		ret = adjust_shift_l2_hdr(iface, buf, eth_hdr);
+		if (ret < 0)
+			return -1;
```

Comment:
An ARP reply always goes back to the same interface that receives it. So we can drop `adjust_shift_l2_hdr()` here, and add a comment explaining why we don't need it. Notice that if we reached this point of execution, Gatekeeper has already checked that the packet has the appropriate Ethernet/VLAN header.

Reply:
OK, so we are adding the assumption that if an interface is configured to insert a VLAN tag on egress packets, all ingress packets will also have a VLAN header?

I'm OK with this assumption. But if this is the case, when we re-use packets, we need to overwrite the VLAN tag with our own, right?

Reply:
When an interface has a VLAN ID, all its ingress packets have a VLAN header **during normal operations**. We should deal with the cases in which ingress packets do have a VLAN ID; these cases can be an attack, a misconfiguration, a network failure, etc. So we shouldn't assume that all ingress packets have a VLAN header. But we can assume a stronger version of this assumption: when an interface has a VLAN ID, all ingress packets that the LLS will reply must have a VLAN ID. This assumption essentially assumes that if a neighbor host, or Gatekeeper itself, is not correctly configured, the host is not a logical neighbor.

The VLAN tag should be the same going in and out of Gatekeeper. As long as we can describe in a comment how we expect that the tag has been checked, we don't need to write it. Notice that this behavior is assumed because we are working with a single interface to receive and transmit the packets. When a packet cross between interfaces, we just update the VLAN ID if the outgoing interface requires a VLAN header.

Reply:
OK, so the LLS block is a special case in this regard, so we can remove `adjust_shift_l2_hdr()`. Got it.

Regarding checking the VLAN tag: I think this comment from above confused me: "Notice that if we reached this point of execution, Gatekeeper has already checked that the packet has the appropriate Ethernet/VLAN header." But the code does not check to make sure the VLAN tag matches. Therefore, I think we need to write the correct tag when re-using the packet.

Reply:
Yes, you got it right. The LLS block is a special case.

If the packet does not have the same VLAN ID, we should not even reply it; logging a warning is appropriate in this case. Trace the paths that a packet goes through to arrive at this point. There's a chance that the VLAN ID is already checked. If so, we just justify in a comment where it's checked for not updating it.

Reply:
By the way, if you address the single comment I have over patch **acl: allow software ntuple filtering**, I can review it again while you work on this one.